airbyte-integrations/bases/debezium-v1-9-6/src/main/java/io/airbyte/integrations/debezium/internals/DebeziumRecordIterator.java

```diff
@@ -71,34 +83,68 @@ protected ChangeEvent<String, String> computeNext() {
       final ChangeEvent<String, String> next;
       try {
-        final Duration waitTime = receivedFirstRecord ? SUBSEQUENT_RECORD_WAIT_TIME : firstRecordWaitTime;
+        // #18987: waitTime here is largely meaningless with heartbeats but leaving for backward
+        // compatibility
+        // for connectors not implementing heartbeat yet (MySql, MSSql)
+        final Duration waitTime = receivedFirstRecord ? SUBSEQUENT_RECORD_WAIT_TIME : this.firstRecordWaitTime;
         next = queue.poll(waitTime.getSeconds(), TimeUnit.SECONDS);
       } catch (final InterruptedException e) {
         throw new RuntimeException(e);
       }
 
       // if within the timeout, the consumer could not get a record, it is time to tell the producer to
       // shutdown.
+      // #18987: Noticed in testing that it's possible for DBZ to be stuck "Searching for WAL resume
+      // position"
+      // when no changes exist. In that case queue will pop after timeout with null value for next
       if (next == null) {
-        LOGGER.info("Closing cause next is returned as null");
+        LOGGER.info("Closing: queue returned null event");
         requestClose();
         LOGGER.info("no record found. polling again.");
         continue;
       }
 
+      // check if heartbeat and read hearbeat position
+      LOGGER.debug("checking heartbeat lsn for: {}", next);
+      final Long heartbeatPos = targetPosition.getHeartbeatPosition(next);
+      if (heartbeatPos != null) {
+        // wrap up sync if heartbeat position crossed the target OR heartbeat position hasn't changed for
+        // too long
+        if (targetPosition.reachedTargetPosition(heartbeatPos)
+            || (heartbeatPos.equals(this.lastHeartbeatPosition) && heartbeatPosNotChanging())) {
+          LOGGER.info("Closing: Heartbeat indicates sync is done");
+          requestClose();
+        }
+        if (!heartbeatPos.equals(this.lastHeartbeatPosition)) {
+          this.tsLastHeartbeat = LocalDateTime.now();
+          this.lastHeartbeatPosition = heartbeatPos;
+        }
+        continue;
+      }
+
       final JsonNode eventAsJson = Jsons.deserialize(next.value());
       hasSnapshotFinished = hasSnapshotFinished(eventAsJson);
 
       // if the last record matches the target file position, it is time to tell the producer to shutdown.
+
       if (!signalledClose && shouldSignalClose(eventAsJson)) {
+        LOGGER.info("Closing: Change event reached target position");
         requestClose();
       }
+      this.tsLastHeartbeat = null;
+      this.lastHeartbeatPosition = null;
       receivedFirstRecord = true;
       return next;
     }
     return endOfData();
   }
 
+  private boolean heartbeatPosNotChanging() {
+    final Duration tbt = Duration.between(this.tsLastHeartbeat, LocalDateTime.now());
+    LOGGER.debug("Time since last hb_pos change {}s", tbt.toSeconds());
+    return tbt.compareTo(SUBSEQUENT_RECORD_WAIT_TIME) > 0;
```

Comment:
this makes me suspect that in case when the source DB is very slow to respond due to compounding issues (slow network, slow disk, overloaded database, who-knows-what...) we are still at the mercy of the hard-coded 1M subsequent record wait time. If that's so - we should make this parameter configurable in spec.

Reply:
Heartbeats are generated locally inside the engine according to configured `heartbeat.interval.ms`. 
Initially I made this timeout use the configured timeout mainly for simplicity but the configured one is typically longer than I think we need to wait. So I opted for a set 1min.
the value we're looking at corresponds with the file offset we are reading the logs from. 
Do you think it's feasible that nothing will move over this time? Maybe if we increase the value?

The reason I'm reluctant to add another timeout is that I don't know how we can explain this value the user in a way that can even be somewhat understood.

Reply:
@grishick Maybe put this at 1/2 of configured wait time? this will give us a lever in case we need, without making it too complicated